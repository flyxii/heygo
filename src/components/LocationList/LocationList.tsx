/**
 * Component to display the input field and the list of locations
 * @props {query} the location query
 */

import React, { useState } from "react";
import InputField from "./InputField";
import List from "./List";

const LocationList: React.FC = () => {
  const [query, setQuery] = useState<string>("");

  return (
    <>
      <InputField
        placeholder="Enter your query here"
        value={query}
        onChangeHandler={(val) => {
          setQuery(val);
        }}
      />
      <List query={query} />
    </>
  );
};

export default LocationList;
