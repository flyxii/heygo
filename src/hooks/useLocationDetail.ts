/**
 * The query to fetch the location detail
 */

import { useQuery, QueryObserverResult } from "react-query";

type LocationDetail = {
  name: string;
  id: number;
  latitude: number;
  longitude: number;
  message?: string;
};

const getLocationDetail = async (id: string) => {
  const response = await fetch(
    `https://code-challenge-backend.herokuapp.com/locations/${id}`
  );
  const data = await response.json();
  return data;
};

export const useLocationDetail = (
  id: string
): QueryObserverResult<LocationDetail, unknown> => {
  return useQuery(["locationDetail", id], () => getLocationDetail(id));
};
