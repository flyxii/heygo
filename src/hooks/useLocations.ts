/**
 * The query to fetch the location list
 */

import { useQuery, QueryObserverResult } from "react-query";

type Locations = {
  name: string;
  id: number;
};

const getLocations = async (query: string) => {
  const response = await fetch(
    `https://code-challenge-backend.herokuapp.com/locations?q=${query}`
  );
  const data = await response.json();
  return data;
};

export const useLocations = (
  query: string
): QueryObserverResult<Locations[], unknown> => {
  return useQuery(["locations", query], () => getLocations(query));
};
