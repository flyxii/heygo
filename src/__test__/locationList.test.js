import * as React from "react";
import { render } from "@testing-library/react";
import LocationList from "components/LocationList";
import { useLocations } from "hooks/useLocations";

const mockedUseLocations = useLocations; 
jest.mock("hooks/useLocations");

describe("<LocationList />", () => {
  beforeEach(() => {
    mockedUseLocations.mockImplementation(() => ({
      isLoading: true,
    }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("Renders without crashing", () => {
    const { getByTestId, asFragment } = render(<LocationList />);

    const inputField = getByTestId("input-field");
    const loading = getByTestId("loading");

    expect(inputField).toBeInTheDocument();
    expect(loading).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("Display error message correctly", () => {
    mockedUseLocations.mockImplementation(() => ({
      isLoading: false,
      error: true,
    }));

    const { getByTestId, asFragment } = render(<LocationList />);

    const error = getByTestId("error");

    expect(error).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("Display error message when no data return", () => {
    const mockData = [];
    mockedUseLocations.mockImplementation(() => ({
      isLoading: false,
      data: mockData,
    }));

    const { getByTestId, asFragment } = render(<LocationList />);

    const error = getByTestId("message-no-data");

    expect(error).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("Displays data correctly", () => {
    const mockData = [{
      name: "mockName",
      id: 1234,
    }];
    mockedUseLocations.mockImplementation(() => ({
      isLoading: false,
      data: mockData,
    }));

    const { getByText, asFragment } = render(<LocationList />);

    const location = getByText("mockName");

    expect(location).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
