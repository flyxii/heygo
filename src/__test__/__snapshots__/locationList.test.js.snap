// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationList /> Display error message correctly 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
    data-testid="input-field"
  >
    <input
      class="inputField"
      placeholder="Enter your query here"
      type="text"
      value=""
    />
  </div>
  <div
    class="wrapper"
    data-testid="error"
  >
    Error fetching data...
  </div>
</DocumentFragment>
`;

exports[`<LocationList /> Display error message when no data return 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
    data-testid="input-field"
  >
    <input
      class="inputField"
      placeholder="Enter your query here"
      type="text"
      value=""
    />
  </div>
  <div
    class="wrapper"
    data-testid="message-no-data"
  >
    No relevant locations could be found
  </div>
</DocumentFragment>
`;

exports[`<LocationList /> Displays data correctly 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
    data-testid="input-field"
  >
    <input
      class="inputField"
      placeholder="Enter your query here"
      type="text"
      value=""
    />
  </div>
  <div
    class="wrapper"
    data-testid="location-list"
  >
    <div
      class="locationList"
    >
      <div
        class="location"
      >
        mockName
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`<LocationList /> Renders without crashing 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
    data-testid="input-field"
  >
    <input
      class="inputField"
      placeholder="Enter your query here"
      type="text"
      value=""
    />
  </div>
  <div
    class="wrapper"
    data-testid="loading"
  >
    We are working hard to get the data...
  </div>
</DocumentFragment>
`;
